sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/RequestResponseIO.java

```diff
@@ -88,73 +355,316 @@ static <RequestT, ResponseT> Builder<RequestT, ResponseT> builder() {
+      PCollection<RequestT> input,
+      PCollectionList<ResponseT> responseList,
+      PCollectionList<ApiIOError> failureList) {
+
+    if (rrioConfiguration.getCacheWrite() == null) {
+      Call<RequestT, ResponseT> call = Call.of(callConfiguration);
+      Result<ResponseT> result = input.apply(CALL_NAME, call);
+      return Pair.of(
+          responseList.and(result.getResponses()), failureList.and(result.getFailures()));
+    }
+
+    // Wrap caller to associate RequestT with ResponseT as a KV.
+    Caller<RequestT, KV<RequestT, ResponseT>> caller =
+        new WrappedAssociatingRequestResponseCaller<>(callConfiguration.getCaller());
+
+    Coder<KV<RequestT, ResponseT>> coder =
+        KvCoder.of(input.getCoder(), rrioConfiguration.getResponseTCoder());
+
+    // Could not re-use original configuration because of different type parameters.
+    Call.Configuration<RequestT, KV<RequestT, ResponseT>> configuration =
+        Call.Configuration.<RequestT, KV<RequestT, ResponseT>>builder()
+            .setResponseCoder(coder)
+            .setCaller(caller)
+            .setSetupTeardown(callConfiguration.getSetupTeardown())
+            .setBackOffSupplier(callConfiguration.getBackOffSupplier())
+            .setCallShouldBackoff(
+                new WrappedAssociatingRequestResponseCallShouldBackoff<>(
+                    callConfiguration.getCallShouldBackoff()))
+            .setShouldRepeat(callConfiguration.getShouldRepeat())
+            .setSleeperSupplier(callConfiguration.getSleeperSupplier())
+            .setTimeout(callConfiguration.getTimeout())
+            .build();
+
+    Call<RequestT, KV<RequestT, ResponseT>> call = Call.of(configuration);
+
+    // Extract ResponseT from KV; append failures.
+    Result<KV<RequestT, ResponseT>> result = input.apply(CALL_NAME, call);
+    PCollection<ResponseT> responses =
+        result.getResponses().apply(CALL_NAME + "Responses", Values.create());
+    responseList = responseList.and(responses);
+    failureList = failureList.and(result.getFailures());
+
+    // Write RequestT and ResponseT pairs to the cache; append failures.
+    Result<KV<RequestT, ResponseT>> cacheWriteResult =
+        result
+            .getResponses()
+            .apply(CACHE_WRITE_NAME, checkStateNotNull(rrioConfiguration.getCacheWrite()));
+    failureList = failureList.and(cacheWriteResult.getFailures());
+
+    return Pair.of(responseList, failureList);
+  }
+
+  private static class PartitionCacheReadsFn<RequestT, ResponseT>
+      implements PartitionFn<KV<RequestT, ResponseT>> {
+    private static final int NUM_PARTITIONS = 2;
+
+    private static final int NON_NULL_PARTITION = 0;
+    private static final int NULL_PARTITION = 1;
+
+    @Override
+    public int partitionFor(KV<RequestT, ResponseT> elem, int numPartitions) {
+      if (checkStateNotNull(elem).getValue() != null) {
+        return NON_NULL_PARTITION;
+      }
+      return NULL_PARTITION;
+    }
+  }
```

Comment:
@johnjcasey or @Abacn Thank you.

Reply:
could you add some documentation on this? its a bit confusing as to what the purpose of this function is. Also, why do we have a constant NUM_PARTITIONS that doesn't appear to be used?